### PR TITLE
Criteria query

### PR DESCRIPTION
**Pull request by DevIgork**

### CODE REVIEW

File: src/main/java/com/example/bookstore/repository/impl/book/TitleSpecificationProvider.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import com.example.bookstore.repository.SpecificationProvider;
+import java.util.Arrays;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "title";
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/repository/impl/book/TitleSpecificationProvider.java]: please avoid String literals

File: src/main/java/com/example/bookstore/repository/impl/book/AuthorSpecificationProvider.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import com.example.bookstore.repository.SpecificationProvider;
+import java.util.Arrays;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "author";
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/repository/impl/book/AuthorSpecificationProvider.java]: please avoid String literals

**Review by d1sam — CHANGES_REQUESTED** [review]
Well done, but please avoid String literals

File: src/main/java/com/example/bookstore/repository/impl/book/TitleSpecificationProvider.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    private static final String KEY = "title";
+
+    @Override
+    public String getKey() {
+        return KEY;
+    }
+
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder)
+                -> root.get("title").in(Arrays.stream(params).toArray());
```
Comment (kshuryhin) [inline src/main/java/com/example/bookstore/repository/impl/book/TitleSpecificationProvider.java]: pay more attention, use constant here

File: src/main/java/com/example/bookstore/repository/impl/book/AuthorSpecificationProvider.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    private static final String KEY = "author";
+
+    @Override
+    public String getKey() {
+        return KEY;
+    }
+
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder)
+                -> root.get("author").in(Arrays.stream(params).toArray());
```
Comment (kshuryhin) [inline src/main/java/com/example/bookstore/repository/impl/book/AuthorSpecificationProvider.java]: pay more attention, use constant here

File: src/main/java/com/example/bookstore/repository/impl/BookSpecificationBuilder.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+                            .getSpecification(searchParameters.title())
+            );
+        }
+        if (searchParameters.author() != null
+                && searchParameters.author().length > 0) {
+            spec = spec.and(
+                    bookSpecificationProviderManager.getSpecificationProvider("author")
+                            .getSpecification(searchParameters.author())
+            );
+        }
```
Comment (kshuryhin) [inline src/main/java/com/example/bookstore/repository/impl/BookSpecificationBuilder.java]: Could we split it to separate methods for better code readability?

**Review by kshuryhin — CHANGES_REQUESTED** [review]
Good job, but there are some points to consider

**Review by vovasalyha — APPROVED** [review]
LGTM

**Pull request merged**